### ISSUE TITLE
 Implement segmenting of Line2D for better fidelity when using width curves

### ISSUE THREAD

**Issue by chryan** [godotengine/godot-proposals#10452]
### Describe the project you are working on

Top-down game

### Describe the problem or limitation you are having in your project

Currently, Line2D doesn't render the correct curved geometry that would intuitively expectat from how define you width curve - it doesn't accommodate well for when the number of points on your line are less than the number of points on your curve.

Here's an example of the problem described:
![line2d-not-showing-ridges](https://github.com/user-attachments/assets/921d52e0-7e40-47df-bbbd-2b66640d075b)

And this is what I expect it to look like:
![line2d-proper-curve-width](https://github.com/user-attachments/assets/1cf6159d-98b5-4873-b706-fcc991465141)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I have a pull request here that fixes it by generating additional points on a line based on the number of segments derived from the curve bake resolution: https://github.com/godotengine/godot/pull/95541

Along with the problem described above, my pull request also adds an extra `curve_offset` property which allows you to do some neat Line2D animations with the curve width (see example videos below).

Segment produce better 'wavey' lines when more curve control points exist than line points:
https://github.com/user-attachments/assets/35a1fe1a-bd1f-4313-bb29-d7d7adb75b57

Curve offset adjusts the sampling position of the curve:
https://github.com/user-attachments/assets/d66792df-39c9-4945-93ab-47e96d49cfcd

You can also animate the curve offset to produce some neat line effects:
https://github.com/user-attachments/assets/e1e2d108-ad7d-4fdd-83b9-3629abf76da2

Other Notes:
- It's possible for the logic to move into the LineBuilder if it's deemed that that's the better place to implement this feature. I'm just not currently familiar with the math behind how that works so I opted for the simpler solution
- The number of segments corresponds with the curve bake resolution instead of a being separate property - I didn't see it being very useful to keep those values independent

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

See video above.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No. This has to be implemented as part of Line2D before the line is built and sent to the rendering server.

### Is there a reason why this should be core and not an add-on in the asset library?

It's a pretty simple change that gives you the visual result that you generally expect when using the curve width feature.
The curve offset also allows for some neat freedom of expression that should really just come standard with Line2D.

**Comment by Calinou** [comment]
My concern with this is that it'll significantly increase the number of curve points without the user realizing it, which has a performance impact. This is particularly the case on integrated graphics and mobile.

If this is to be implemented, it should be something you manually opt into, so that you only pay the performance cost when you need it.

**Comment by kleonc** [comment]
Also note there's the same issue with Line2D's gradient, see: https://github.com/godotengine/godot/issues/50076. It might make sense to tackle both issues at once.

**Comment by chryan** [comment]
> My concern with this is that it'll significantly increase the number of curve points without the user realizing it, which has a performance impact. This is particularly the case on integrated graphics and mobile.
> 
> If this is to be implemented, it should be something you manually opt into, so that you only pay the performance cost when you need it.

I'm aware of and agree with you about the potential performance implications of this change.

The main reasons for choosing this approach are **_correctness and ease-of-use_** which I posit supersedes performance implications that can come from the excessive use of this Line2Ds with curve widths.

I would argue that using the curve width in its current incarnation is broken - if the points on your line don't match up with the points on your curve, it's not apparent why your line doesn't render according to the visual representation of your curve.
In fact, it wasn't apparent to me why my line wasn't rendering the way the feature presented itself until I started to randomly throw in points on the curve, and subsequently digging into the C++ engine code. Adding additional properties further complicates the number of dials you have to tune to get your line rendering in a way that a default user's intuition presents (the gradient that @kleonc brought up also illustrates this complication).

But to play devil's advocate, let's breakdown the potential performance implications of this:
The default bake resolution for curves is 100, yielding approximately 100 segments on your line. With each segment being made up of 2 triangles, you get about 200 triangles per curve-width-line (my algorithm doesn't result in exact number of segments equal to the bake resolution).
Is a default 200 triangles for a single curve-width-line really that much of a performance concern even on integrated graphics and mobile? Even in the **_very bespoke_** use-case of using a 100+ of these curve-width-lines yields triangles in the range of ~20000, which is pretty standard for a single 3d mesh these days, even on integrated cards and mobile devices.

To me, a complete, easy-to-use, working feature is more important than the theoretical performance loss in very specific use-cases of games that need to render 100s or even 1000s of a Line2D. I would assume that developers leveraging Line2D to that extent would already need to be aware that a performance hit would need to be addressed on their end.

To highlight your legitimate concerns, I propose that the solution for this is user education via additional tooltip information on the curve_width property (i.e. increasing the bake_resolution affects the number of segments which can blow up your tri-count/performance).

**Comment by chryan** [comment]
> Also note there's the same issue with Line2D's gradient, see: [godotengine/godot#50076](https://github.com/godotengine/godot/issues/50076). It might make sense to tackle both issues at once.

I can work on an implementation for this as well.

**Comment by Calinou** [comment]
> Is a default 200 triangles for a single curve-width-line really that much of a performance concern even on integrated graphics and mobile?

Yes, because the 200 triangles need to be generated *every frame* if the Line2D is animated. This is where most of the performance cost comes from.

**Comment by chryan** [comment]
Thanks for the feedback.
I did some profiling to verify the performance claims.

100 segments, 1000 animating lines:
![line2d-lots-of-segments](https://github.com/user-attachments/assets/eeacd0ab-3ac1-45b7-ba90-b21ebb9e0ab0)

16 segments, 1000 animating lines:
![line2d-less-line-segments](https://github.com/user-attachments/assets/d39d4164-2bb8-49dc-b4b8-2d18fdb202ff)

I've added a new `_curve_line_segments` property that defaults to 1, but the resulting segments used will be always be higher (details below).

I've also gone ahead and implemented a fix for issue https://github.com/godotengine/godot/issues/50076.
My implementation inserts points along the total line that correspond with the normalize 0 to 1 offset value of each point on the gradient (and excludes matching line points):
https://github.com/user-attachments/assets/ff759192-9c7b-4fc9-9e42-043427ea8e46

In order to fix the gradients, points have to be inserted separate from the usage of the curve segments feature, so if a gradient is defined, the number of base line segments goes up.

The minimum number of segments will always be at least the number of points on the (optionally) gradient-point-inclusive line.

Here's a video to show the fidelity increase as the segments property goes up:
https://github.com/user-attachments/assets/bb60220d-9879-4981-badd-0311765a8961

I've verified that the line works with all the different capping methods as well.

**Comment by NathanLovato** [comment]
I second the goal of the proposal to offer options to make the line draw as intended by the user, and having a property to control the amount of subdivisions sounds great. We have similar controls on 3D mesh tools like CSGPolygon3D when drawing along a path.

CSG meshes can be much costlier performance-wise due to the third dimension plus boolean operations, but the point is offering tools to prototype and iterate fast. I see the same value in Line2D. Being an all-purpose drawing tool for lines and curves, it's never going to be super efficient.

In practice, currently, for prototyping we tend to generate points in GDScript, which results in lower performance than a native option would, and then usually ditch Line2D because of its limitations and use our own shaders and/or drawing code as needed to e.g. pan a texture along a a mesh designed and unwrapped in Blender, curving it at runtime instead of generating points.

**Comment by QbieShay** [comment]
I feel like having subdivisions sounds great, but the offset option I'm a little more hesitant with. This is something that can be done in a shader, even if with more difficulty. I'm not opposed to it, but I'm wondering at what point we stop to say "ok no, this feature should be a shader feature not a line feature".

**Comment by NathanLovato** [comment]
It's probably up to the maintainers in charge of this area to join the discussion and decide on the purpose and scope of the feature and what may go in or not.
